File: src/M20_sdk_deploy/scripts/lidar_to_scan.py
#!/usr/bin/env python3
"""半球 LiDAR → policy obs (与 sim/rl_training/sensors/patterns.py 一致)

输入: base_link 系下的 PointCloud2 (Robosense Airy)
输出: Float32MultiArray (992 维) = 前向 16×31 (496) + 后向 16×31 (496)
布局: row-major (polar 外, azimuth 内)
"""
import math
import time

import numpy as np
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import PointCloud2
from sensor_msgs_py import point_cloud2
from std_msgs.msg import Float32MultiArray


NUM_POLAR = 16
NUM_AZ = 31
NUM_PER_DIR = NUM_POLAR * NUM_AZ          # 496
NUM_TOTAL = NUM_PER_DIR * 2                # 992
MAX_DIST = 2.5
MIN_DIST = 0.3
HALF_PI = math.pi / 2
TWO_PI = 2 * math.pi


class LidarToScanArrayNode(Node):
    def __init__(self):
        super().__init__('lidar_to_scan_array_node')
        self.subscription = self.create_subscription(
            PointCloud2, '/LIDAR_SIM_RAW', self.pc_callback, 10)
        self.array_pub = self.create_publisher(
            Float32MultiArray, '/scan/multi_layer_features_array', 10)

        self.fwd_offset = np.array([0.32028, 0.0, -0.013], dtype=np.float32)
        self.bwd_offset = np.array([-0.32028, 0.0, -0.013], dtype=np.float32)

        self.recv_count = 0
        self.pub_count = 0
        self.process_times = []
        self.diag_timer = self.create_timer(1.0, self.log_diagnostics)

    @staticmethod
    def _bin_one_direction(pts, boresight_sign):
        """boresight_sign = +1: 前向 (boresight=+X); -1: 后向 (boresight=-X, 配合 Y 取反)"""
        bins = np.full(NUM_PER_DIR, MAX_DIST, dtype=np.float32)
        if pts.size == 0:
            return bins

        x = pts[:, 0]
        y = pts[:, 1]
        z = pts[:, 2]

        if boresight_sign > 0:
            mask = x > 0.0
            x_b = x
            y_b = y
        else:
            mask = x < 0.0
            x_b = -x
            y_b = -y                               # 180°Z 旋转 → Y 取反

        if not np.any(mask):
            return bins
        x_b = x_b[mask]; y_b = y_b[mask]; z_b = z[mask]

        r = np.sqrt(x_b * x_b + y_b * y_b + z_b * z_b)
        valid = (r >= MIN_DIST) & (r <= MAX_DIST)
        if not np.any(valid):
            return bins
        x_b = x_b[valid]; y_b = y_b[valid]; z_b = z_b[valid]; r = r[valid]

        perp = np.sqrt(y_b * y_b + z_b * z_b)
        polar = np.arctan2(perp, x_b)              # [0, π/2]
        az = np.arctan2(y_b, z_b)                  # [-π, π]

        p_idx = np.clip(np.rint(polar * (NUM_POLAR - 1) / HALF_PI).astype(np.int32),
                        0, NUM_POLAR - 1)
        a_idx = np.clip(np.rint((az + math.pi) * (NUM_AZ - 1) / TWO_PI).astype(np.int32),
                        0, NUM_AZ - 1)
        flat = p_idx * NUM_AZ + a_idx

        np.minimum.at(bins, flat, r)
        return bins

    def pc_callback(self, msg):
        self.recv_count += 1
        start_time = time.time()

        pt_data = point_cloud2.read_points(msg, field_names=("x", "y", "z"), skip_nans=True)
        pt_list = list(pt_data)
        if not pt_list:
            return

        tmp_arr = np.array(pt_list)
        if tmp_arr.dtype.names is not None:
            points = np.column_stack((tmp_arr['x'], tmp_arr['y'], tmp_arr['z'])).astype(np.float32)
        else:
            points = tmp_arr.astype(np.float32)

        if len(points) == 0:
            return

        rel_fwd = points - self.fwd_offset
        rel_bwd = points - self.bwd_offset

        fwd_bins = self._bin_one_direction(rel_fwd, +1)
        bwd_bins = self._bin_one_direction(rel_bwd, -1)

        # 兜底盲区 (与 sim multi_layer_scan 一致：< 0.3m → 2.5m)
        fwd_bins[fwd_bins < MIN_DIST] = MAX_DIST
        bwd_bins[bwd_bins < MIN_DIST] = MAX_DIST

        array_msg = Float32MultiArray()
        array_msg.data = fwd_bins.tolist() + bwd_bins.tolist()
        self.array_pub.publish(array_msg)

        self.pub_count += 1
        self.process_times.append((time.time() - start_time) * 1000.0)

    def log_diagnostics(self):
        if self.recv_count > 0:
            avg_time = sum(self.process_times) / len(self.process_times) if self.process_times else 0.0
            color = "\033[92m" if self.recv_count >= 8 else "\033[93m"
            reset = "\033[0m"
            self.get_logger().info(
                f"📊 [Scan Node] {color}Recv: {self.recv_count} Hz | Pub: {self.pub_count} Hz{reset} | "
                f"Avg Process Time: {avg_time:.2f} ms"
            )
        else:
            self.get_logger().warn("⚠️ [Scan Node] Waiting for PointCloud2 data on /LIDAR_SIM_RAW ...")
        self.recv_count = 0
        self.pub_count = 0
        self.process_times.clear()


def main(args=None):
    rclpy.init(args=args)
    node = LidarToScanArrayNode()
    node.get_logger().info(
        f"✅ Lidar to Scan Array Node started. Hemispherical {NUM_POLAR}×{NUM_AZ} per dir, "
        f"output dim = {NUM_TOTAL}.")
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/M20_sdk_deploy/src/lidar_to_scan.cpp
// 半球 LiDAR → policy obs 转换 (Sim 端 HemisphericalLidarPatternCfg 对齐)
//
// 输入：base_link 系下的 PointCloud2 (Robosense Airy)
// 输出：Float32MultiArray (992 维)
//   [0   .. 495]  前向半球 16 polar × 31 azimuth (row-major: polar 外, azimuth 内)
//   [496 .. 991]  后向半球，同样 16 × 31 layout
//
// 几何 (与 sim/rl_training/sensors/patterns.py 完全一致):
//   - polar  ∈ [0, POLAR_MAX_RAD = 80°]:  从 boresight 出发的极角
//        - 真 Airy Lissajous 扫描密度在 polar > 84° 急剧下降，导致 sim2real
//          gap (real polar=90° no-hit 86% vs sim 59%)。把 sim 与 deploy
//          都收紧到 80°，确保两端密度匹配。polar > 80° 的真机点直接丢弃。
//   - azimuth ∈ [-π, π]:    绕 boresight 一圈，从 +Z 起算，正向 +Y
//   - 前 boresight = robot +X，后 boresight = robot -X (对应 sim 后 sensor 的 180°Z 旋转)
//
// Bin 规则 (nearest neighbor, 匹配 sim 的 linspace 采样点):
//   polar_idx   = round(polar   * (NUM_POLAR-1)   / POLAR_MAX_RAD)  ∈ [0, 15]
//   azimuth_idx = round((azimuth + π) * (NUM_AZ-1) / (2π))           ∈ [0, 30]
//   flat_idx    = polar_idx * NUM_AZ + azimuth_idx
//
// 每 bin 取最小深度；< 0.3m 视为盲区 → 填 2.5m (no-hit max_distance, 与 sim 一致)。

#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/point_cloud2.hpp>
#include <sensor_msgs/point_cloud2_iterator.hpp>
#include <std_msgs/msg/float32_multi_array.hpp>
#include <cmath>
#include <vector>
#include <chrono>
#include <algorithm>

class LidarToScanArrayNode : public rclcpp::Node {
public:
    LidarToScanArrayNode() : Node("lidar_to_scan_array_node") {
        this->declare_parameter<std::string>("lidar_topic", "/LIDAR_SIM_RAW");
        std::string lidar_topic = this->get_parameter("lidar_topic").as_string();

        sub_ = this->create_subscription<sensor_msgs::msg::PointCloud2>(
            lidar_topic, rclcpp::SensorDataQoS(),
            std::bind(&LidarToScanArrayNode::pc_callback, this, std::placeholders::_1));

        pub_ = this->create_publisher<std_msgs::msg::Float32MultiArray>(
            "/scan/multi_layer_features_array", 10);

        diag_timer_ = this->create_wall_timer(
            std::chrono::seconds(1),
            std::bind(&LidarToScanArrayNode::log_diagnostics, this));

        RCLCPP_INFO(this->get_logger(),
            "✅ [Scan C++] Hemispherical 16×31 binning. Subscribing to: %s. Output dim: %d.",
            lidar_topic.c_str(), NUM_TOTAL);
    }

private:
    static constexpr int NUM_POLAR   = 16;
    static constexpr int NUM_AZ      = 31;
    static constexpr int NUM_PER_DIR = NUM_POLAR * NUM_AZ;   // 496
    static constexpr int NUM_TOTAL   = NUM_PER_DIR * 2;       // 992

    static constexpr float MAX_DIST       = 2.5f;
    static constexpr float MIN_DIST       = 0.3f;
    // 真 Airy 在 polar > 80° 数据极不可靠，sim 与 deploy 同步限制到 80°
    static constexpr float POLAR_MAX_DEG  = 80.0f;
    static constexpr float POLAR_MAX_RAD  = POLAR_MAX_DEG * static_cast<float>(M_PI) / 180.0f;
    static constexpr float TWO_PI         = static_cast<float>(2.0 * M_PI);

    void pc_callback(const sensor_msgs::msg::PointCloud2::SharedPtr msg) {
        recv_count_++;
        auto start_time = std::chrono::high_resolution_clock::now();

        std::vector<float> fwd_bins(NUM_PER_DIR, MAX_DIST);
        std::vector<float> bwd_bins(NUM_PER_DIR, MAX_DIST);

        sensor_msgs::PointCloud2ConstIterator<float> iter_x(*msg, "x");
        sensor_msgs::PointCloud2ConstIterator<float> iter_y(*msg, "y");
        sensor_msgs::PointCloud2ConstIterator<float> iter_z(*msg, "z");

        for (; iter_x != iter_x.end(); ++iter_x, ++iter_y, ++iter_z) {
            const float px = *iter_x;
            const float py = *iter_y;
            const float pz = *iter_z;
            if (!std::isfinite(px) || !std::isfinite(py) || !std::isfinite(pz)) continue;

            // ===== 前向半球 (boresight = +X) =====
            {
                const float fx = px - fwd_offset_[0];
                const float fy = py - fwd_offset_[1];
                const float fz = pz - fwd_offset_[2];
                if (fx > 0.0f) {
                    const float r = std::sqrt(fx*fx + fy*fy + fz*fz);
                    if (r >= MIN_DIST && r <= MAX_DIST) {
                        const float perp  = std::sqrt(fy*fy + fz*fz);
                        const float polar = std::atan2(perp, fx);          // [0, π/2]
                        if (polar <= POLAR_MAX_RAD) {
                            const float az = std::atan2(fy, fz);            // [-π, π]
                            const int p_idx = clampi(
                                static_cast<int>(std::lround(polar * (NUM_POLAR - 1) / POLAR_MAX_RAD)),
                                0, NUM_POLAR - 1);
                            const int a_idx = clampi(
                                static_cast<int>(std::lround((az + static_cast<float>(M_PI)) * (NUM_AZ - 1) / TWO_PI)),
                                0, NUM_AZ - 1);
                            const int idx = p_idx * NUM_AZ + a_idx;
                            if (r < fwd_bins[idx]) fwd_bins[idx] = r;
                        }
                    }
                }
            }

            // ===== 后向半球 (boresight = -X)，对应 sim 后 sensor 的 180°Z 旋转 =====
            {
                const float bx = px - bwd_offset_[0];
                const float by = py - bwd_offset_[1];
                const float bz = pz - bwd_offset_[2];
                if (bx < 0.0f) {
                    const float r = std::sqrt(bx*bx + by*by + bz*bz);
                    if (r >= MIN_DIST && r <= MAX_DIST) {
                        const float perp  = std::sqrt(by*by + bz*bz);
                        const float polar = std::atan2(perp, -bx);          // 从 -X 算
                        if (polar <= POLAR_MAX_RAD) {
                            const float az = std::atan2(-by, bz);            // Y 取反 (180°Z mirror)
                            const int p_idx = clampi(
                                static_cast<int>(std::lround(polar * (NUM_POLAR - 1) / POLAR_MAX_RAD)),
                                0, NUM_POLAR - 1);
                            const int a_idx = clampi(
                                static_cast<int>(std::lround((az + static_cast<float>(M_PI)) * (NUM_AZ - 1) / TWO_PI)),
                                0, NUM_AZ - 1);
                            const int idx = p_idx * NUM_AZ + a_idx;
                            if (r < bwd_bins[idx]) bwd_bins[idx] = r;
                        }
                    }
                }
            }
        }

        // 盲区处理：< 0.3m 视为 no-hit (与 sim multi_layer_scan 一致)
        for (auto& v : fwd_bins) { if (v < MIN_DIST) v = MAX_DIST; }
        for (auto& v : bwd_bins) { if (v < MIN_DIST) v = MAX_DIST; }

        std_msgs::msg::Float32MultiArray array_msg;
        array_msg.data.reserve(NUM_TOTAL);
        array_msg.data.insert(array_msg.data.end(), fwd_bins.begin(), fwd_bins.end());
        array_msg.data.insert(array_msg.data.end(), bwd_bins.begin(), bwd_bins.end());
        pub_->publish(array_msg);

        pub_count_++;
        auto end_time = std::chrono::high_resolution_clock::now();
        total_time_ms_ += std::chrono::duration<double, std::milli>(end_time - start_time).count();
    }

    static inline int clampi(int v, int lo, int hi) {
        return v < lo ? lo : (v > hi ? hi : v);
    }

    void log_diagnostics() {
        if (recv_count_ > 0) {
            double avg_time = total_time_ms_ / recv_count_;
            RCLCPP_INFO(this->get_logger(),
                "📊 [Scan C++] Recv: %d Hz | Pub: %d Hz | Avg Process Time: %.3f ms",
                recv_count_, pub_count_, avg_time);
        } else {
            RCLCPP_WARN(this->get_logger(), "⚠️ [Scan C++] Waiting for PointCloud2...");
        }
        recv_count_ = 0;
        pub_count_ = 0;
        total_time_ms_ = 0.0;
    }

    rclcpp::Subscription<sensor_msgs::msg::PointCloud2>::SharedPtr sub_;
    rclcpp::Publisher<std_msgs::msg::Float32MultiArray>::SharedPtr pub_;
    rclcpp::TimerBase::SharedPtr diag_timer_;

    const float fwd_offset_[3] = {0.32028f, 0.0f, -0.013f};
    const float bwd_offset_[3] = {-0.32028f, 0.0f, -0.013f};

    int recv_count_ = 0;
    int pub_count_ = 0;
    double total_time_ms_ = 0.0;
};

int main(int argc, char * argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<LidarToScanArrayNode>());
    rclcpp::shutdown();
    return 0;
}
